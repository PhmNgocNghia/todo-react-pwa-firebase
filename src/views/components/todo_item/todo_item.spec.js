
import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import Button from '../button.js'
import Checkbox from '../checkbox'
import TodoItem from './index';
import TodoItemName from '../todo_item_name'

const defaultProp = {
  todo: {
    name: 'test',
    completed: false,
    key: '123'
  }
}

describe('Todo Item', function() {
  // Render
  it('should render input, checkbox delete, edit button when not in edit mode', function () {
    const shallowTodoItem = shallow(<TodoItem {...defaultProp}/>)
    const Buttons = shallowTodoItem.find(Button)
    expect(shallowTodoItem.find(TodoItemName)).toHaveLength(1)
    expect(shallowTodoItem.find(Checkbox)).toHaveLength(1)
    expect(Buttons).toHaveLength(2)
    expect(Buttons.find({
      icon:"edit",
      text:"Update"
    })).toHaveLength(1)
    expect(Buttons.find({
      icon:"trash",
      text:"Delete"
    })).toHaveLength(1)
  })

  it('should render input, change, go back button when not in edit mode', function () {
    const shallowTodoItem = shallow(<TodoItem {...defaultProp}/>)
    shallowTodoItem.find('.btn-update').simulate('click')
    const Buttons = shallowTodoItem.find(Button)
    expect(shallowTodoItem.find('input')).toHaveLength(1)
    expect(Buttons).toHaveLength(2)
    expect(Buttons.find({
      icon:"edit",
      text:"Change"
    })).toHaveLength(1)
    expect(Buttons.find({
      icon:"window-close",
      text:"Go back"
    })).toHaveLength(1)
  })
  // State
  it('should go edit mode when press edit button', function () {
    const shallowTodoItem = shallow(<TodoItem {...defaultProp}/>)
    shallowTodoItem.find('.btn-update').simulate('click')
    expect(shallowTodoItem.state('isEditMode')).toBeTruthy()
  })

  it('should go back when press edit button', function () {
    const shallowTodoItem = shallow(<TodoItem {...defaultProp}/>)
    shallowTodoItem.find('.btn-update').simulate('click')
    shallowTodoItem.find('.btn-goBack').simulate('click')
    expect(shallowTodoItem.state('isEditMode')).toBeFalsy()
  })

  // Mock function
  it('should call delete function when press delete button', function () {
    const mockFunctionProps = {
      ...defaultProp,
      onRemoveTodo: jest.fn()
    }
    const shallowTodoItem = shallow(<TodoItem {...mockFunctionProps}/>)
    shallowTodoItem.find('.btn-delete').simulate('click')
    expect(mockFunctionProps.onRemoveTodo).toHaveBeenCalled()
    expect(mockFunctionProps.onRemoveTodo).toHaveBeenCalledWith(defaultProp.todo)
  })

  it('should not call update function when press change button and text is not change', () => {
    const mockFunctionProps = {
      ...defaultProp,
      onUpdateTodo: jest.fn()
    }
    const shallowTodoItem = shallow(<TodoItem {...mockFunctionProps}/>)
    shallowTodoItem.find('.btn-update').simulate('click')
    shallowTodoItem.find('.btn-change').simulate('click')
    expect(mockFunctionProps.onUpdateTodo).toHaveBeenCalledTimes(0)
  })

  it('should call update function when press change button and text is not empty', () => {
    const mockFunctionProps = {
      ...defaultProp,
      onUpdateTodo: jest.fn()
    }
    const shallowTodoItem = shallow(<TodoItem {...mockFunctionProps}/>)
    shallowTodoItem.find('.btn-update').simulate('click')
    shallowTodoItem.find('input').simulate('change', {
      target: {value: 'newVal'}
    })
    shallowTodoItem.find('.btn-change').simulate('click')
    expect(mockFunctionProps.onUpdateTodo).toHaveBeenCalled()
    expect(mockFunctionProps.onUpdateTodo).toHaveBeenCalledWith({
      name: 'newVal',
      completed: false,
      key: '123'
    })
  })

  it('should call remove function when press change button and text empty', () => {
    const mockFunctionProps = {
      ...defaultProp,
      onRemoveTodo: jest.fn()
    }
    const shallowTodoItem = shallow(<TodoItem {...mockFunctionProps}/>)
    shallowTodoItem.find('.btn-update').simulate('click')
    shallowTodoItem.setState({
      newTodoName: ''
    })
    shallowTodoItem.find('.btn-change').simulate('click')
    expect(mockFunctionProps.onRemoveTodo).toHaveBeenCalled()
    expect(mockFunctionProps.onRemoveTodo).toHaveBeenCalledWith(defaultProp.todo)
  })

  it('should toggle todo completed true when click checkbox', () => {
    const mockFunctionProps = {
      ...defaultProp,
      onUpdateTodo: jest.fn()
    }

    const shallowTodoItem = shallow(<TodoItem {...mockFunctionProps}/>)
    const shallowCheckbox = shallowTodoItem.find(Checkbox).dive()
    shallowCheckbox.find('input').simulate('change')
    expect(mockFunctionProps.onUpdateTodo).toHaveBeenCalled()
    expect(mockFunctionProps.onUpdateTodo).toHaveBeenCalledWith({
      ...defaultProp.todo,
      completed: true
    })
  })

  it('should toggle todo completed false when click checkbox', () => {
    const mockFunctionProps = {
      todo: {
        ...defaultProp.todo,
        completed: true
      },
      onUpdateTodo: jest.fn()
    }

    const shallowTodoItem = shallow(<TodoItem {...mockFunctionProps}/>)
    const shallowCheckbox = shallowTodoItem.find(Checkbox).dive()
    shallowCheckbox.find('input').simulate('change')
    expect(mockFunctionProps.onUpdateTodo).toHaveBeenCalled()
    expect(mockFunctionProps.onUpdateTodo).toHaveBeenCalledWith({
      ...defaultProp.todo,
      completed: false
    })
  })

  // Snapshoot
  it('should match snap shoot when not in edit mode', function () {
    const shallowTodoItem = shallow(<TodoItem {...defaultProp}/>)
    let Json = toJson(shallowTodoItem)
    expect(Json).toMatchSnapshot()
  })

  it('should match snap shoot when in edit mode', function () {
    const shallowTodoItem = shallow(<TodoItem {...defaultProp}/>)
    shallowTodoItem.setState({
      isEditMode: true
    })
    let Json = toJson(shallowTodoItem)
    expect(Json).toMatchSnapshot()
  })
})