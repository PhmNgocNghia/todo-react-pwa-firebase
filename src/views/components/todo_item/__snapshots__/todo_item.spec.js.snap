// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Item should match snap shoot when in edit mode 1`] = `
<styled.div
  className="pt-3 pb-1 d-flex align-items-center"
>
  <input
    className="todo_item_input form-control pl-0"
    key="pug:0:0"
    onChange={[Function]}
    onKeyUp={[Function]}
    value=""
  />
  <button
    className="btn-change"
    icon="edit"
    key="pug:0:1"
    onClick={[Function]}
    text="Change"
  />
  <button
    className="btn-goBack"
    icon="window-close"
    key="pug:0:2"
    onClick={[Function]}
    text="Go back"
  />
</styled.div>
`;

exports[`Todo Item should match snap shoot when not in edit mode 1`] = `
<styled.div
  className="pt-3 pb-1 d-flex align-items-center"
>
  <Checkbox
    checked={false}
    id="123"
    key="pug:1:0"
    onCheck={[Function]}
  />
  <styled.button
    className="todo_name mr-auto"
    completed={false}
    key="pug:1:1"
    onClick={[Function]}
  >
    test
  </styled.button>
  <button
    className="btn-update"
    icon="edit"
    key="pug:1:2"
    onClick={[Function]}
    text="Update"
  />
  <button
    className="btn-delete"
    icon="trash"
    key="pug:1:3"
    onClick={[Function]}
    text="Delete"
  />
</styled.div>
`;
